src/plugins/azure/tunnel.ts

```diff
@@ -0,0 +1,123 @@
+
+const TUNNEL_READY_STRING = "Tunnel is ready";
+const SPAWN_TUNNEL_TRIES = 3;
+
+export type BastionTunnelMeta = {
+  killTunnel: () => void;
+  tunnelLocalPort: string;
+};
+
+const selectRandomPort = (): string => {
+  // The IANA ephemeral port range is 49152 to 65535, inclusive. Pick a random value in that range.
+  // If the port is in use (unlikely but possible), we can just generate a new value and try again.
+  // 16384 is 65535 - 49152 + 1, the number of possible ports in the range.
+  const port = Math.floor(Math.random() * 16384) + 49152;
+  return port.toString();
+};
+
+const spawnBastionTunnelInBackground = (
+  request: AzureSshRequest,
+  port: string
+): Promise<BastionTunnelMeta> => {
+  return new Promise<BastionTunnelMeta>((resolve, reject) => {
+    let processTerminated = false;
+    let tunnelReady = false;
+    let stdout = "";
+    let stderr = "";
+
+    const child = spawn(
+      "az",
+      [
+        "network",
+        "bastion",
+        "tunnel",
+        "--ids",
+        request.bastionId,
+        "--target-resource-id",
+        request.instanceId,
+        "--resource-port",
+        "22",
+        "--port",
+        port,
+      ],
+      // Spawn the process in detached mode so that it is in its own process group; this lets us kill it and all
+      // descendent processes together.
+      { detached: true }
+    );
+
+    child.on("exit", (code) => {
+      if (tunnelReady) return;
+
+      // We don't expect the process to terminate on its own before the tunnel is ready
+      print2(stdout);
+      print2(stderr);
+      reject(
+        `Unable to start Azure Network Bastion tunnel; tunnel process ended with status ${code}`
+      );
+    });
+
+    child.stdout.on("data", (data) => {
+      const str = data.toString("utf-8");
+      stdout += str;
+    });
+
+    child.stderr.on("data", (data) => {
+      const str = data.toString("utf-8");
+      stderr += str;
+
+      if (str.includes(TUNNEL_READY_STRING)) {
+        print2("Azure Bastion tunnel is ready.");
+        tunnelReady = true;
+
+        resolve({
+          killTunnel: () => {
+            if (processTerminated) return;
+
+            processTerminated = true;
+
+            if (child.pid) {
+              // Kill the process and all its descendents via killing the process group; this is only possible
+              // because we launched the process with `detached: true` above. This is necessary because `az` is
+              // actually a bash script that spawns a Python process, and we need to kill the Python process as well.
+              // SIGINT is equivalent to pressing Ctrl-C in the terminal; allows for the tunnel process to perform any
+              // necessary cleanup of its own before exiting. The negative PID is what indicates that we want to kill
+              // the whole process group.
+              process.kill(-child.pid, "SIGINT");
+            }
+
+            child.unref();
```

Comment:
The only reason I have the `if (child.pid)` check is because `child.pid` is typed as `number | undefined`, when it being `undefined` if the process failed to exec. `child` itself should never be undefined, even if the process fails to spawn, and `unref()` should just do nothing in that case.